src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs

```diff
@@ -102,13 +103,42 @@ protected override void DisconnectHandler(WKWebView nativeView)
 			Services != null;
 
 		private string? HostPage { get; set; }
-		private ObservableCollection<RootComponent>? RootComponents { get; set; }
-		private new IServiceProvider? Services { get; set; }
+		private ObservableCollection<RootComponent>? RootComponents
+		{
+			get => _rootComponents;
+			set
+			{
+				if (_rootComponents != null)
+				{
+					// Remove any previously-known root components and unhook events
+					_rootComponents.Clear();
+					_rootComponents.CollectionChanged -= OnRootComponentsCollectionChanged;
+				}
+
+				_rootComponents = value;
+
+				if (_rootComponents != null)
+				{
+					// Add new root components and hook events
+					if (_rootComponents.Count > 0 && _webviewManager != null)
+					{
+						_ = _webviewManager.Dispatcher.InvokeAsync(async () =>
+						{
+							foreach (var component in _rootComponents)
+							{
+								await component.AddToWebViewManagerAsync(_webviewManager);
+							}
+						});
+					}
+					_rootComponents.CollectionChanged += OnRootComponentsCollectionChanged;
```

Comment:
Also race here (or again, maybe not, depending on whether we're already on the dispatcher)